cumulus/pallets/weight-reclaim/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+			"Proof recording enabled during prepare, now disabled. This should not happen.",
+		) else {
+			return Ok(inner_refund)
+		};
+
+		// The consumed proof size as measured by the host.
+		let measured_proof_size =
+			proof_size_after_dispatch.saturating_sub(proof_size_before_dispatch);
+
+		// The consumed weight as benchmarked. Calculated from post info and info.
+		// NOTE: `calc_actual_weight` will take the minimum of `post_info` and `info` weights.
+		// This means any underestimation of compute time in the pre dispatch info will not be
+		// taken into account.
+		let benchmarked_actual_weight = post_info_with_inner.calc_actual_weight(info);
+
+		let benchmarked_actual_proof_size = benchmarked_actual_weight.proof_size();
+		if benchmarked_actual_proof_size < measured_proof_size {
+			log::error!(
+				target: LOG_TARGET,
+				"Benchmarked storage weight smaller than consumed storage weight. \
+				benchmarked: {benchmarked_actual_proof_size} consumed: {measured_proof_size}"
+			);
+		} else {
+			log::trace!(
+				target: LOG_TARGET,
+				"Reclaiming storage weight. benchmarked: {benchmarked_actual_proof_size},
+				consumed: {measured_proof_size}"
+			);
+		}
+
+		let accurate_weight = benchmarked_actual_weight.set_proof_size(measured_proof_size);
+		frame_system::BlockWeight::<T>::mutate(|current_weight| {
+			let already_reclaimed = frame_system::ExtrinsicWeightReclaimed::<T>::get();
+			current_weight.accrue(already_reclaimed, info.class);
+			current_weight.reduce(info.total_weight(), info.class);
+			current_weight.accrue(accurate_weight, info.class);
+
+			// If we encounter a situation where the node-side proof size is already higher than
+			// what we have in the runtime bookkeeping, we add the difference to the `BlockWeight`.
+			// This prevents that the proof size grows faster than the runtime proof size.
+			let extrinsic_len = frame_system::AllExtrinsicsLen::<T>::get().unwrap_or(0);
+			let node_side_pov_size = proof_size_after_dispatch.saturating_add(extrinsic_len.into());
+			let block_weight_proof_size = current_weight.total().proof_size();
+			let missing_from_node = node_side_pov_size.saturating_sub(block_weight_proof_size);
+			if missing_from_node > 0 {
+				log::warn!(
+					target: LOG_TARGET,
+					"Node-side PoV size higher than runtime proof size weight. node-side: \
+					{node_side_pov_size} extrinsic_len: {extrinsic_len} runtime: \
+					{block_weight_proof_size}, missing: {missing_from_node}. Setting to node-side \
+					proof size."
+				);
+				current_weight.accrue(Weight::from_parts(0, missing_from_node), info.class);
+			}
+		});
+
+		// The saturation will happen if the pre dispatch weight is underestimating the proof
+		// size.
+		// In this case the extrinsic proof size weight reclaimed is 0.
+		let accurate_unspent = info.total_weight().saturating_sub(accurate_weight);
+		frame_system::ExtrinsicWeightReclaimed::<T>::put(accurate_unspent);
```

Comment:
You are right, it would be more meaningful to include this information into the reclaimed amount. I did in https://github.com/paritytech/polkadot-sdk/pull/6140/commits/3c8c0e7c7158f19724db5aad6bd185830606b656